### ISSUE TITLE
MacOS: Packing rn's bundle always stuck with Recrawled

### ISSUE THREAD

**Issue by PTLEI** [facebook/watchman#981]
I'm facing many 'Recrawled' problems recently.They stuck me in the packing process.The total process takes me an hour and a half what took me just several minutes before.
I have solved several problems with raising the default per-process descriptor limit and updating the ruby version(other packages with the same updating).
But I have the last problem with the info of 'FSEvents flag', I have no idea to resolve it
`
Recrawled this watch 1 times, most recently because:
MustScanSubDirs UserDroppedTo resolve, please review the information on
https://facebook.github.io/watchman/docs/troubleshooting.html#recrawl
To clear this warning, run:
 'watchman watch-del '/Users/lanleilei/temp/mobile' ; watchman watch-project '/Users/lanleilei/temp/mobile' '
`

MacOS: 11.6 Big Sur
ruby: 2.7.5p203
watchman: 2021.12.06.00

**Comment by kmancini** [comment]
Hello, sorry to hear about these Watchman troubles. 

Can you confirm that the 'watchman watch-del '/Users/lanleilei/temp/mobile' ; watchman watch-project '/Users/lanleilei/temp/mobile' commands this error suggested did not resolve this issue? If so could you link your watchman logs? Typically, the logs should be in the same directory where the watchman socket is which you can find with `watchman get-sockname`.

**Comment by PTLEI** [comment]
> 

First I not sure.I just use Watchman to bundle the react-native app. These abnormal logs are printed where the process stuck, so I try to resolve them.
Yesterday I upgraded my os to 'Monterey 12.1' and try to bundle again this morning.The process still stunk, but it just took 18 minutes which down from an hour and a half
Yesterday logs (maybe): 
`
2021-12-16T19:36:54,619: [listener] waiting for 1 clients to terminate
2021-12-16T19:36:55,139: [sanitychecks] done with sanityCheckThread
2021-12-16T19:36:58,266: [listener] 3 roots were still live at exit
2021-12-16T19:36:58,266: [listener] Exiting from service with res=true
2021-12-16T19:56:25,087: [listener] Watchman 2021.12.06.00 <no build info set> starting up on lanleileideMacBook-Pro.local
2021-12-16T19:56:25,089: [listener] path /Users/lanleilei/temp/mobile is on filesystem type apfs
2021-12-16T19:56:25,091: [listener] failed to use watcher kqueue+fsevents: Not using the kqueue+fsevents watcher as the "prefer_split_fsevents_watcher" config isn't set.
2021-12-16T19:56:25,091: [listener] root /Users/lanleilei/temp/mobile using watcher mechanism fsevents (auto was requested)
2021-12-16T19:56:25,093: [listener] path /Users/lanleilei/mobile_dev/mobile is on filesystem type apfs
2021-12-16T19:56:25,095: [listener] failed to use watcher kqueue+fsevents: Not using the kqueue+fsevents watcher as the "prefer_split_fsevents_watcher" config isn't set.
2021-12-16T19:56:25,095: [listener] root /Users/lanleilei/mobile_dev/mobile using watcher mechanism fsevents (auto was requested)
2021-12-16T19:56:25,096: [listener] file limit is 2560 kern.maxfilesperproc=61440
2021-12-16T19:56:25,096: [listener] raised file limit to 61440
`
Today (early):
`
2021-12-17T09:53:53,757: [listener] 2 roots were still live at exit
2021-12-17T09:53:53,757: [listener] Exiting from service with res=true
2021-12-17T09:53:53,810: [listener] Watchman 2021.12.06.00 <no build info set> starting up on lanleileideMacBook-Pro.local
2021-12-17T09:53:53,811: [listener] path /Users/lanleilei/temp/mobile is on filesystem type apfs
2021-12-17T09:53:53,812: [listener] failed to use watcher kqueue+fsevents: Not using the kqueue+fsevents watcher as the "prefer_split_fsevents_watcher" config isn't set.
2021-12-17T09:53:53,813: [listener] root /Users/lanleilei/temp/mobile using watcher mechanism fsevents (auto was requested)
2021-12-17T09:53:53,814: [listener] path /Users/lanleilei/mobile_dev/mobile is on filesystem type apfs
2021-12-17T09:53:53,814: [listener] failed to use watcher kqueue+fsevents: Not using the kqueue+fsevents watcher as the "prefer_split_fsevents_watcher" config isn't set.
2021-12-17T09:53:53,814: [listener] root /Users/lanleilei/mobile_dev/mobile using watcher mechanism fsevents (auto was requested)
2021-12-17T09:53:53,815: [listener] file limit is 2560 kern.maxfilesperproc=61440
2021-12-17T09:53:53,815: [listener] raised file limit to 61440
2021-12-17T09:53:53,815: [listener] launchd: "sock" wasn't present in Sockets
2021-12-17T09:53:53,816: [sanitychecks] starting sanityCheckThread
2021-12-17T09:53:55,374: [io 140625618947432 /Users/lanleilei/mobile_dev/mobile] PERF: {"system_time": 1.6714830000000001, "user_time": 0.49976199999999998, "elapsed_time": 1.559215, "pid": 6451, "meta": {"root": {"watcher": "fsevents", "case_sensitive": false, "recrawl_count": 0, "path": "/Users/lanleilei/mobile_dev/mobile"}}, "start_time": 1639706033.8153181, "version": "2021.12.06.00", "description": "full-crawl"}
2021-12-17T09:53:55,374: [client=2:stm=105553123198976:pid=3557] PERF: {"system_time": 1.6647889999999999, "user_time": 0.49687599999999998, "elapsed_time": 1.5521989999999999, "pid": 6451, "meta": {"args": ["watch-project", "/Users/lanleilei/mobile_dev/mobile"], "client": {"pid": 3557}, "root": {"watcher": "fsevents", "case_sensitive": false, "recrawl_count": 0, "path": "/Users/lanleilei/mobile_dev/mobile"}}, "start_time": 1639706033.8223701, "version": "2021.12.06.00", "description": "dispatch_command:watch-project"}
2021-12-17T09:53:55,375: [io 140625618947432 /Users/lanleilei/mobile_dev/mobile] crawl complete
2021-12-17T09:53:58,179: [io 140625621043416 /Users/lanleilei/temp/mobile] PERF: {"system_time": 3.4202409999999999, "user_time": 0.97314199999999995, "elapsed_time": 4.3655520000000001, "pid": 6451, "meta": {"root": {"watcher": "fsevents", "case_sensitive": false, "recrawl_count": 0, "path": "/Users/lanleilei/temp/mobile"}}, "start_time": 1639706033.814158, "version": "2021.12.06.00", "description": "full-crawl"}
2021-12-17T09:53:58,179: [io 140625621043416 /Users/lanleilei/temp/mobile] crawl complete
`
After I upgraded Watchman to '2021.12.13.00' (used 18mins):
`
2021-12-17T10:00:56,666: [fsevents /Users/lanleilei/mobile_dev/mobile] PERF: {"system_time": 1.0000000000000001e-5, "user_time": 5.1999999999999997e-5, "elapsed_time": 1.1e-5, "pid": 6451, "meta": {"root": {"watcher": "fsevents", "case_sensitive": false, "recrawl_count": 0, "path": "/Users/lanleilei/mobile_dev/mobile"}}, "start_time": 1639706456.6662259, "version": "2021.12.06.00", "description": "UserDropped"}
2021-12-17T10:00:56,666: [notify 140625618947432 /Users/lanleilei/mobile_dev/mobile] /Users/lanleilei/mobile_dev/mobile: MustScanSubDirs UserDropped: scheduling a tree recrawl
2021-12-17T10:01:01,437: [io 140625618947432 /Users/lanleilei/mobile_dev/mobile] PERF: {"system_time": 2.7168099999999997, "user_time": 1.1888749999999999, "elapsed_time": 4.7707579999999998, "pid": 6451, "meta": {"root": {"watcher": "fsevents", "case_sensitive": false, "recrawl_count": 2, "path": "/Users/lanleilei/mobile_dev/mobile"}}, "start_time": 1639706456.6671231, "version": "2021.12.06.00", "description": "full-crawl"}
2021-12-17T10:01:01,437: [io 140625618947432 /Users/lanleilei/mobile_dev/mobile] recrawl complete
2021-12-17T10:02:02,183: [client=12:stm=105553123382272:pid=22329] PERF: {"system_time": 5.272716, "user_time": 2.4315549999999999, "elapsed_time": 1.4000539999999999, "pid": 6451, "meta": {"args": ["query", "/Users/lanleilei/mobile_dev/mobile", {"glob": ["**/*.js", "**/*.json", "**/*.ts", "**/*.tsx", "**/*.bmp", "**/*.gif", "**/*.jpg", "**/*.jpeg", "**/*.png", "**/*.psd", "**/*.svg", "**/*.webp", "**/*.m4v", "**/*.mov", "**/*.mp4", "**/*.mpeg", "**/*.mpg", "**/*.webm", "**/*.aac", "**/*.aiff", "**/*.caf", "**/*.m4a", "**/*.mp3", "**/*.wav", "**/*.html", "**/*.json", "**/*.pdf", "**/*.yaml", "**/*.yml", "**/*.otf", "**/*.ttf", "**/*.zip"], "fields": ["name", "exists", "mtime_ms", "size", "content.sha1hex"], "expression": ["allof", ["type", "f"], ["anyof", ["suffix", "js"], ["suffix", "json"], ["suffix", "ts"], ["suffix", "tsx"], ["suffix", "bmp"], ["suffix", "gif"], ["suffix", "jpg"], ["suffix", "jpeg"], ["suffix", "png"], ["suffix", "psd"], ["suffix", "svg"], ["suffix", "webp"], ["suffix", "m4v"], ["suffix", "mov"], ["suffix", "mp4"], ["suffix", "mpeg"], ["suffix", "mpg"], ["suffix", "webm"], ["suffix", "aac"], ["suffix", "aiff"], ["suffix", "caf"], ["suffix", "m4a"], ["suffix", "mp3"], ["suffix", "wav"], ["suffix", "html"], ["suffix", "json"], ["suffix", "pdf"], ["suffix", "yaml"], ["suffix", "yml"], ["suffix", "otf"], ["suffix", "ttf"], ["suffix", "zip"]]]}], "client": {"pid": 22329}, "root": {"watcher": "fsevents", "case_sensitive": false, "recrawl_count": 2, "path": "/Users/lanleilei/mobile_dev/mobile"}}, "start_time": 1639706520.7831421, "version": "2021.12.06.00", "description": "dispatch_command:query"}
2021-12-17T10:02:57,601: [client=15:stm=105553189271040:pid=6451] PERF: {"system_time": 0.0020339999999999998, "user_time": 0.00076400000000000003, "elapsed_time": 1.000181, "pid": 6451, "meta": {"args": ["clock", "/Users/lanleilei/temp/mobile", {"sync_timeout": 20000}], "client": {"pid": 6451}, "root": {"watcher": "fsevents", "case_sensitive": false, "recrawl_count": 0, "path": "/Users/lanleilei/temp/mobile"}}, "start_time": 1639706576.601609, "version": "2021.12.06.00", "description": "dispatch_command:clock"}
`

**Comment by sergtimosh** [comment]
I'm having same issue with my project

`Recrawled this watch 31 times, most recently because:
MustScanSubDirs UserDroppedTo resolve, please review the information on
https://facebook.github.io/watchman/docs/troubleshooting.html#recrawl
To clear this warning, run:
watchman watch-del '/Users/serhiytymoshenko/work/proj/app_master' ; watchman watch-project '/Users/serhiytymoshenko/work/proj/app_master'`

This probably the reason why hot reload not working for me now.

MacOS: 12.1 Monterey
ruby: 2.6.8p205
watchman: 2021.12.20.00

p.s. I have colleagues with older watchman  versions without this issues. Their versions are **2021.10.18.00**  and **2021.09.06.00**

**Comment by raarts** [comment]
Exactly the same issue here @sergtimosh is reporting.

**Comment by curthipster** [comment]
Same here, initially on v2021.11.29.00, and also after upgrading to v2022.01.17.00.

Log sample:
```
2022-01-19T15:26:01,103: [notify 4762770664 /Users/chrispruett/dev/jam-prod] /Users/chrispruett/dev/jam-prod: MustScanSubDirs UserDropped: scheduling a tree recrawl
2022-01-19T15:26:01,537: [io 4762770664 /Users/chrispruett/dev/jam-prod] PERF: {"system_time": 0.31507200000000002, "user_time": 0.073803999999999995, "elapsed_time": 0.43441400000000002, "pid": 8574, "meta": {"root": {"watcher": "fsevents", "case_sensitive": false, "recrawl_count": 4, "path": "/Users/chrispruett/dev/jam-prod"}}, "start_time": 1642634761.1032741, "version": "2021.11.29.00", "description": "full-crawl"}
```

**Comment by jason0x43** [comment]
I have also been experiencing this issue for the last month or so.

```
watchman warning:  Recrawled this watch 35 times, most recently because:
MustScanSubDirs UserDroppedTo resolve, please review the information on
https://facebook.github.io/watchman/docs/troubleshooting.html#recrawl
To clear this warning, run: ...
```

I've seen that warning in one standard React Native project and one Expo project.

I got an M1 Mac at around the same time the issue started happening, so I'm not sure if the issue is related to the new architecture or to the fact that the version of watchman that was released around then is problematic.

**Comment by liamjones** [comment]
> I got an M1 Mac at around the same time the issue started happening, so I'm not sure if the issue is related to the new architecture or to the fact that the version of watchman that was released around then is problematic.

I suspect it's related to a watchman change in recent releases, I'm seeing it on an Intel Mac here.

**Comment by jason0x43** [comment]
This is happening almost every time I start a dev server for any of my React Native projects. I did try downgrading to 2021.11.29 , which did not appear to exhibit this problem. Unfortunately there are no release notes for watchman, and most of the updates lately seem to have come from submodules, so it's not terribly easy to see what might have changed.

**Comment by christianrank** [comment]
For me this started happening since I switched to a MacBook with M1 chip. Never have seen this before.

**Comment by kdenz** [comment]
Same issue here, really annoying

**Comment by jimmihagenTrifecta** [comment]
I am also seeing 

> 'Recrawled this watch 21 times, most recently because:
> MustScanSubDirs UserDroppedTo resolve, please review the information on
> https://facebook.github.io/watchman/docs/troubleshooting.html#recrawl'

 after moving development to an M1 Macbook

**Comment by wassimAJ** [comment]
Same issue here, I'm using M1 Macbook.

**Comment by MoOx** [comment]
For people facing this on a new machine, it's just because you didn't adjusted some OS settings.

If you follow carefully the link attached to the warning and effectively try to correct the issue, the message will go away.

This link https://facebook.github.io/watchman/docs/troubleshooting.html#recrawl mention this one https://facebook.github.io/watchman/docs/install.html#system-specific-preparation under the section "Avoiding Recrawls", which gives a hint.

For example on macOS, just do this in your terminal to raise the limit has explained.

```console
echo "kern.maxfiles=10485760\nkern.maxfilesperproc=1048576\n" | sudo tee /etc/sysctl.conf
```

Then reboot your mac and you should be good to go.

**Comment by sergtimoshUptarget** [comment]
> https://facebook.github.io/watchman/docs/install.html#system-specific-preparation

The manual says _Only applicable on macOS 10.6 and earlier_.
Just tried it on M1 Monterey, the is still here.

**Comment by sergtimosh** [comment]
> https://facebook.github.io/watchman/docs/install.html#system-specific-preparation

The manual says _Only applicable on macOS 10.6 and earlier_.
Just tried it on M1 Monterey, the issue still here.

upd. This approach also wasn't helpful https://superuser.com/questions/1634286/how-do-i-increase-the-max-open-files-in-macos-big-sur

**Comment by MoOx** [comment]
@sergtimosh that's right. See https://github.com/facebook/watchman/issues/880#issuecomment-749687420 for confirmation. In theory nothing is required for macOS 10.7+

**Comment by jason0x43** [comment]
This has been mentioned before, but I'm also seeing this on an Intel Mac (at least with v2022.01.03), so it's not M1-specific.

**Comment by MoOx** [comment]
Probably just related to be moving to a new machine, migrated or not.

**Comment by sergtimosh** [comment]
@MoOx   I've been setting up two different machines with M1 from scratch(not migrating stuff) - both had same issue with watchman described in this thread.

**Comment by PTLEI** [comment]
@MoOx In my case, it just happened in upgrading Catalina to Big Sur. I find that they have huge different in 'File Descriptor Limits', Big Sur's limit is far less than the recommended values which the official given. But it hard to modify those limits, they back to default as I reboot again. Finally I upgrade to Monterey and endure about 10 mins' waiting

**Comment by jason0x43** [comment]
> Probably just related to be moving to a new machine, migrated or not.

Entirely possible -- I started seeing it after setting up a new M1 Mac in November, and I've seen it a few times on an Intel Mac I setup last month.

**Comment by sergtimosh** [comment]
@jason0x43

>  and I've seen it a few times on an Intel Mac I setup last month.

 then it's related to the latest watchman versions. As I mentioned earlier the MacBook I was setting up first doesn't had this issue(versions 2021.10.18.00 and 2021.09.06.00). The one  I started to work with on December (watchman: 2021.12.20.00) was spitting this _Recrawled_ into the output.

**Comment by rodrigodiasf1984** [comment]
watchman warning:  Recrawled this watch 45 times, most recently because:
MustScanSubDirs UserDroppedTo resolve, please review the information on
https://facebook.github.io/watchman/docs/troubleshooting.html#recrawl
To clear this warning, run:
`watchman watch-del '/Users/rodrigodiasdefigueiredo/Desktop/test-app' ; watchman watch-project '/Users/rodrigodiasdefigueiredo/Desktop/test-app`

I'm using a MacBook air M1, react-native. 0.67.2. and MacOs Monterey

**Comment by ryparker** [comment]
This is happening to me as well when running jest tests on a NextJS project (not react native):

```sh
$ yarn test
yarn run v1.22.17
$ jest
watchman warning:  Recrawled this watch 27 times, most recently because:
MustScanSubDirs UserDroppedTo resolve, please review the information on
https://facebook.github.io/watchman/docs/troubleshooting.html#recrawl
To clear this warning, run:
`watchman watch-del '/Users/ryanparker/GitHub/parliamentarian' ; watchman watch-project '/Users/ryanparker/GitHub/parliamentarian'`
```

Clearing the warning does not resolve the issue.

Using M1 Max Macbook Pro on Monterey 12.2.1

**Comment by animesh-csmp** [comment]
Yup, it seems to happening when I switched from Intel based MacAir to arm64 based M1Pro. The build doesn't fails, though the warning pops up now and then

**Comment by iAmHarmon** [comment]
Same here. Using MBP w/ M1 Max on Monterey.

**Comment by Harris-Miller** [comment]
Same here. MBP w/ Intel i9 on Big Sur

**Comment by iAmHarmon** [comment]
@Harris-Miller I was able to fix this using the troubleshooting guide on the Watchman docs [here](https://facebook.github.io/watchman/docs/troubleshooting.html#avoiding-recrawls) and [here] (https://facebook.github.io/watchman/docs/install.html#system-specific-preparation).

TL;DR
Run the following commands to increase the per-process descriptor limit:
 `$ sudo sysctl -w kern.maxfiles=10485760
$ sudo sysctl -w kern.maxfilesperproc=1048576`

To persist this change across reboots, create a file at `/etc/sysctl.conf` if it doesn't already exist. Add the following into the file.
`kern.maxfiles=10485760
kern.maxfilesperproc=1048576`

**Comment by MoOx** [comment]
> Only applicable on macOS 10.6 and earlier

**Comment by iAmHarmon** [comment]
@MoOx im on a MBP w/ M1 Max running Monterey. It worked for me. Worst case scenario, if this doesn’t work for other people, they can delete ‘/etc/sysctl.conf’ and reboot their computer to revert the changes.

**Comment by iAmHarmon** [comment]
@Harris-Miller rereading your comment, I somehow glossed over the fact that you’re using an intel-based MBP. Updating your descriptor limits may not work. I would clear the warning and see if it happens again.

**Comment by jason0x43** [comment]
I'd be interested to see if that fix holds up, or if Watchman is just temporarily behaving. Apparently /etc/sysctl.conf doesn't work to persist kqueue modifications on Big Sur or Monterey (you have to use a [launch daemon](https://superuser.com/a/1646927/1137809)), and Watchman on 10.6+ [uses fsevents](https://github.com/facebook/watchman/issues/880#issuecomment-749687420), so the kqueue settings shouldn't affect it.

**Comment by Harris-Miller** [comment]
> @Harris-Miller rereading your comment, I somehow glossed over the fact that you’re using an intel-based MBP. Updating your descriptor limits may not work. I would clear the warning and see if it happens again.

I do clear it and it continues to happen consistently. In the repos I work in the most I have to clear it daily

**Comment by archcorsair** [comment]
I can confirm I'm experiencing this issue with watchman version `2022.03.21.00`
This started when I switched from an intel based mac to arm64
```
watchman 2022.03.21.00
Monterey 12.3
M1 Pro
```

Shutting down the server and reinstalling it via brew did not solve my issue, warnings come back shortly.

**Comment by AndrewSouthpaw** [comment]
> To persist this change across reboots ... [with sysctl]

I tried this and it didn't work for me. I'm on an Intel-based MBP.

**Comment by look4nico** [comment]
`watchman watch-del-all `

This worked for me.
 - macOS Monterey 12.3.1
 - M1 Pro

**Comment by jimmihagenTrifecta** [comment]
> 



> `watchman watch-del-all `
> 
> This worked for me.
> 
> * macOS Monterey 12.3.1
> * M1 Pro

That doesn't work for long, my system is the same...

**Comment by MyPalVikram** [comment]
Any updates on this issue by chance?

**Comment by swizes** [comment]
Macbook Pro 2018 fresh installed has same issue.

**Comment by dariabr** [comment]
Any updates on this issue?

**Comment by Svarto** [comment]
Any update on this, same problem here after migrating to m1 mac

**Comment by algarcia-vector** [comment]
any updates?

**Comment by sibelius** [comment]
it keeps showing this to me


watchman warning:  Recrawled this watch 9 times, most recently because:
MustScanSubDirs UserDroppedTo resolve, please review the information on
https://facebook.github.io/watchman/docs/troubleshooting.html#recrawl
To clear this warning, run:

I've tried many things, but none solved it

my .watchmanconfig  is {}

**Comment by anietieasuquo** [comment]
I had this issue since moving to M1 with Monterey.

And I fixed this by modifying my .watchmanconfig as follows;

`{
    "fsevents_latency": 0.05,
    "fsevents_try_resync": "true",
    "prefer_split_fsevents_watcher": "true"
}`

The default for `fsevents_latency` is `0.01` in later versions of Watchman. By increasing we allow the system to batch more change notifications together.

Then run; `watchman shutdown-server`, and then try running Watchman again.

You can refer to: https://facebook.github.io/watchman/docs/config.html#fsevents-latency

**Comment by sibelius** [comment]
I've got this error when using the config above

All requests will continue to fail with this message until you resolve
the underlying problem.  You will find more information on fixing this at
https://facebook.github.io/watchman/docs/troubleshooting.html#poison-opendir. Make sure watchman is running for this project. See https://facebook.github.io/watchman/docs/troubleshooting

**Comment by AndrewSouthpaw** [comment]
Yeah I got the same

**Comment by uloco** [comment]
> And I fixed this by modifying my .watchmanconfig as follows;
> 
> `{ "fsevents_latency": 0.05, "fsevents_try_resync": "true", "prefer_split_fsevents_watcher": "true" }`
> 

My watchman crashes with this config and I cannot start metro at all.

**Comment by OmarAl150** [comment]
Well, i have the same issue, but my solution to that was reseting the Watchman, like this

https://stackoverflow.com/questions/49443341/watchman-crawl-failed-retrying-once-with-node-crawler

**Comment by xyliusun** [comment]
I fix this with this step: https://stackoverflow.com/a/71779949/7356847

**Comment by AndrewSouthpaw** [comment]
For future Googlers: I fixed the issue eventually by just uninstalling watchman. 🙃 (Obviously only helps people who don't need watchman directly, maybe you installed it incidentally.)  It was coming up because of running Jest, and apparently it doesn't need to be installed for that to run, and was causing issues with Jest.

**Comment by johnslemmer** [comment]
> 
> To persist this change across reboots, create a file at `/etc/sysctl.conf` if it doesn't already exist. Add the following into the file.
> 
> ```
> kern.maxfiles=10485760
> kern.maxfilesperproc=1048576
> ```

The above answer used to be correct to address this issue across reboots.  But `sysctl.conf` doesn't work anymore in recent versions of macOS.  This stack overflow explains how to do it now: https://superuser.com/questions/1634286/how-do-i-increase-the-max-open-files-in-macos-big-sur

**Comment by Hmoulvad** [comment]
What is the status on this? Pleas help us out here.

**Comment by 2017398956** [comment]
https://facebook.github.io/watchman/docs/cookies#limitation-macos-fsevents

**Comment by cfecherolle** [comment]
Bumping this issue, still happening in 2023 on macOS Monterey 12.6, on a MacBook Pro (16", 2019, not M1) when launching. Is there a solution which is compatible with recent macOS versions?

**Comment by yagofrancia** [comment]
Same problem here, using M2 Air Ventura 13.0

**Comment by lchenfox** [comment]
The same issue on `M1` Mac Book Pro.

```
warning: Watchman `watch-project` returned a warning: Recrawled this watch 47 times, most recently because:
MustScanSubDirs UserDroppedTo resolve, please review the information on
https://facebook.github.io/watchman/docs/troubleshooting.html#recrawl
To clear this warning, run:
`watchman watch-del '/Users/langke/Developer/myapp' ; watchman watch-project '/Users/langke/Developer/myapp``

**Comment by emrbli** [comment]
for MacOS M1;
`$ watchman shutdown-server`
`$ brew update`
`$ brew reinstall watchman`

**Comment by cubabit** [comment]
> for MacOS M1 solved for me; `$ watchman shutdown-server` `$ brew update` `$ brew reinstall watchman`
> 
> for more question; https://facebook.github.io/watchman/docs/troubleshooting.html#recrawl

I ran this and brew upgraded watchman to `2023.08.28.00` but I still get the same issue on a MacBook Pro with M1 Max

**Comment by jzaefferer** [comment]
with `2023.10.09.00` I'm also getting this warning.

```
warning: Watchman `query` returned a warning: Recrawled this watch 3 times, most recently because:
MustScanSubDirs UserDroppedTo resolve, please review the information on
https://facebook.github.io/watchman/docs/troubleshooting.html#recrawl
To clear this warning, run:
`watchman watch-del '/Users/joern.zaefferer/dev/mytonies-app' ; watchman watch-project '/Users/joern.zaefferer/dev/mytonies-app'`
```

I can clear the warning, but it will come back pretty quickly.

I've checked https://facebook.github.io/watchman/docs/troubleshooting.html#recrawl and did not find any mention of `MustScanSubDirs UserDroppedTo resolve`.

Any idea what 'MustScanSubDirs UserDroppedTo resolve' means? It seems like this GitHub issue is one of the very few pages that Google can find that mention it at all. 

Searching within this repo, there's only one occurrence of `MustScanSubDirs`: https://github.com/facebook/watchman/blob/28b3a655a2de1b7940ca67440e7489604d199fbb/watchman/watcher/fsevents.cpp#L73

This also shows up in the `fsevents` package, which looks like an unrelated nodejs wrapper around fsevents: https://github.com/search?q=repo%3Afsevents%2Ffsevents%20MustScanSubDirs&type=code They added handling of `MustScanSubDirs` back in April: https://github.com/fsevents/fsevents/commit/a77340f4f96a6e02cd5bfabaaebd4e10535ee533

In Apple's docs I found this page: https://developer.apple.com/documentation/coreservices/1455361-fseventstreameventflags/kfseventstreameventflagmustscansubdirs/

> Your application must rescan not just the directory given in the event, but all its children, recursively. This can happen if there was a problem whereby events were coalesced hierarchically. For example, an event in /Users/jsmith/Music and an event in /Users/jsmith/Pictures might be coalesced into an event with this flag set and path=/Users/jsmith. If this flag is set you may be able to get an idea of whether the bottleneck happened in the kernel (less likely) or in your client (more likely) by checking for the presence of the informational flags kFSEventStreamEventFlagUserDropped or kFSEventStreamEventFlagKernelDropped.

Since the warning mentions both `MustScanSubDirs` and `UserDroppedTo` I guess this means we can `kFSEventStreamEventFlagMustScanSubDirs` and `kFSEventStreamEventFlagUserDropped`. Aka "the bottleneck happened  in your client".

I know way too little about fsevents to do anything with that though.

**Comment by mmerickel** [comment]
This is still an issue with watchman `2024.04.01.00`, and has been an issue for a long time for us. The `watch-del` does fix things for a minute but it comes back almost immediately and is basically a constant problem. Testing currently on latest macos 14.4.1.

I learned several team members working on react native projects have just been uninstalling watchman because it's so annoying.

I also maintain https://github.com/pylons/hupper which supports a watchman backend and it experiences the same warnings/problems. I'd welcome any tips / suggestions to try to solve it, but ultimately it seems to be some issue within watchman itself on macos.

**Comment by TrustDec** [comment]
2024 not any updates?

**Comment by cubabit** [comment]
> 2024 not any updates?

I wonder how much longer this can be ignored considering a large proportion of developers use Macs and the proportion of Macs that use Apple Silicon is growing rapidly.

The silence is deafening. Is this even being worked on?

**Comment by cfecherolle** [comment]
Are there any news on this issue? Still have it on a M2 Macbook Pro with macOS Sonoma 14.5

**Comment by cubabit** [comment]
Maybe they do not use Macs at Facebook....

**Comment by JaberFarahHii** [comment]
**In the immortal words of LoLs 2024 cinematic:**

> _I’m still here. I’m still here. And I’m still here. I’m still._

And so is this issue. Its 2025, and its still happening. Any updates?

**Comment by andrewhamon** [comment]
Its a bit early, but I have seen improvement by setting either of [prefer_split_fsevents_watcher](https://facebook.github.io/watchman/docs/config#prefer_split_fsevents_watcher) or [fsevents_try_resync](https://facebook.github.io/watchman/docs/config#fsevents_try_resync) (I set both of them, but setting either one was enough to make it so I couldn't reproduce my errors).

**Comment by Abdallah-Ben** [comment]
> for MacOS M1 solved for me; `$ watchman shutdown-server` `$ brew update` `$ brew reinstall watchman`
> 
> for more question; https://facebook.github.io/watchman/docs/troubleshooting.html#recrawl

I followed these instructions and I managed to fix the problem, but before executing `$ brew reinstall watchman` i found out that I have some outdated packages, so I run `$ brew upgrade` and then I run `$ brew reinstall watchman`
